File: auth_client/src/AuthClient.java
import java.io.*;
import java.net.*;
import java.util.Scanner;

/**
 * AuthClient - TCP Client for User Authentication Server
 * Connects to Python auth server and provides interactive CLI
 */
public class AuthClient {
    private Socket socket;
    private BufferedReader reader;
    private PrintWriter writer;
    private Scanner scanner;
    private String currentToken;
    private String serverHost;
    private int serverPort;
    private boolean running;
    
    public AuthClient(String host, int port) {
        this.serverHost = host;
        this.serverPort = port;
        this.scanner = new Scanner(System.in);
        this.currentToken = null;
        this.running = true;
    }
    
    /**
     * Connect to the authentication server
     */
    public boolean connect() {
        try {
            socket = new Socket(serverHost, serverPort);
            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            writer = new PrintWriter(socket.getOutputStream(), true);
            System.out.println("Connected to auth server at " + serverHost + ":" + serverPort);
            return true;
        } catch (IOException e) {
            System.err.println("Failed to connect: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Disconnect from server
     */
    public void disconnect() {
        try {
            if (socket != null && !socket.isClosed()) {
                if (currentToken != null) {
                    sendCommand("LOGOUT " + currentToken);
                    readResponse();
                }
                sendCommand("QUIT");
                readResponse();
                socket.close();
            }
        } catch (IOException e) {
            System.err.println("Error during disconnect: " + e.getMessage());
        }
    }
    
    /**
     * Send command to server
     */
    private void sendCommand(String command) {
        writer.println(command);
        System.out.println("[SENT] " + command);
    }
    
    /**
     * Read response from server
     */
    private String readResponse() throws IOException {
        String response = reader.readLine();
        System.out.println("[RECV] " + response);
        return response;
    }
    
    /**
     * Register a new user
     */
    private void register() throws IOException {
        System.out.print("Enter username: ");
        String username = scanner.nextLine().trim();
        System.out.print("Enter password: ");
        String password = scanner.nextLine().trim();
        
        if (username.isEmpty() || password.isEmpty()) {
            System.out.println("Error: Username and password cannot be empty");
            return;
        }
        
        sendCommand("REGISTER " + username + " " + password);
        String response = readResponse();
        
        if (response.equals("OK")) {
            System.out.println("Registration successful!");
        } else {
            System.out.println("Registration failed: " + response);
        }
    }
    
    /**
     * Login with credentials
     */
    private void login() throws IOException {
        System.out.print("Enter username: ");
        String username = scanner.nextLine().trim();
        System.out.print("Enter password: ");
        String password = scanner.nextLine().trim();
        
        if (username.isEmpty() || password.isEmpty()) {
            System.out.println("Error: Username and password cannot be empty");
            return;
        }
        
        sendCommand("LOGIN " + username + " " + password);
        String response = readResponse();
        
        if (response.startsWith("TOKEN:")) {
            currentToken = response.substring(6);
            System.out.println("Login successful!");
            System.out.println("Session token acquired (hidden for security)");
        } else {
            System.out.println("Login failed: " + response);
        }
    }
    
    /**
     * Access a protected resource
     */
    private void accessResource() throws IOException {
        if (currentToken == null) {
            System.out.println("Error: You must login first");
            return;
        }
        
        System.out.print("Enter resource ID: ");
        String resourceId = scanner.nextLine().trim();
        
        if (resourceId.isEmpty()) {
            System.out.println("Error: Resource ID cannot be empty");
            return;
        }
        
        sendCommand("ACCESS " + currentToken + " " + resourceId);
        String response = readResponse();
        
        if (response.startsWith("DATA:")) {
            System.out.println("Access granted: " + response.substring(5));
        } else {
            System.out.println("Access denied: " + response);
        }
    }
    
    /**
     * Check session status
     */
    private void checkStatus() throws IOException {
        if (currentToken == null) {
            System.out.println("Error: You must login first");
            return;
        }
        
        sendCommand("STATUS " + currentToken);
        String response = readResponse();
        
        if (response.startsWith("ACTIVE:")) {
            String[] parts = response.split(":");
            if (parts.length >= 3) {
                System.out.println("Session active for user: " + parts[1]);
                System.out.println("Minutes remaining: " + parts[2]);
            } else {
                System.out.println("Session active");
            }
        } else {
            System.out.println("Session invalid: " + response);
            currentToken = null;
        }
    }
    
    /**
     * Logout current user
     */
    private void logout() throws IOException {
        if (currentToken == null) {
            System.out.println("Error: No active session");
            return;
        }
        
        sendCommand("LOGOUT " + currentToken);
        String response = readResponse();
        
        if (response.equals("OK")) {
            System.out.println("Logout successful!");
            currentToken = null;
        } else {
            System.out.println("Logout failed: " + response);
        }
    }
    
    /**
     * Print help menu
     */
    private void printHelp() {
        System.out.println("\n=== Available Commands ===");
        System.out.println("1. register - Register a new user");
        System.out.println("2. login    - Login with credentials");
        System.out.println("3. access   - Access a protected resource");
        System.out.println("4. status   - Check session status");
        System.out.println("5. logout   - Logout current session");
        System.out.println("6. help     - Show this help menu");
        System.out.println("7. quit     - Exit the client");
        System.out.println("========================\n");
    }
    
    /**
     * Main interactive loop
     */
    public void run() {
        System.out.println("\n=== User Authentication Client ===");
        System.out.println("Connected to server: " + serverHost + ":" + serverPort);
        printHelp();
        
        while (running) {
            try {
                System.out.print("\nauth> ");
                String input = scanner.nextLine().trim().toLowerCase();
                
                switch (input) {
                    case "1":
                    case "register":
                        register();
                        break;
                    case "2":
                    case "login":
                        login();
                        break;
                    case "3":
                    case "access":
                        accessResource();
                        break;
                    case "4":
                    case "status":
                        checkStatus();
                        break;
                    case "5":
                    case "logout":
                        logout();
                        break;
                    case "6":
                    case "help":
                        printHelp();
                        break;
                    case "7":
                    case "quit":
                    case "exit":
                        running = false;
                        break;
                    default:
                        System.out.println("Unknown command: " + input);
                        System.out.println("Type 'help' for available commands");
                        break;
                }
            } catch (IOException e) {
                System.err.println("Connection error: " + e.getMessage());
                System.out.println("Attempting to reconnect...");
                if (!reconnect()) {
                    System.out.println("Failed to reconnect. Exiting...");
                    running = false;
                }
            }
        }
        
        disconnect();
        System.out.println("Goodbye!");
    }
    
    /**
     * Attempt to reconnect to server
     */
    private boolean reconnect() {
        try {
            if (socket != null && !socket.isClosed()) {
                socket.close();
            }
            return connect();
        } catch (IOException e) {
            return false;
        }
    }
    
    /**
     * Main entry point
     */
    public static void main(String[] args) {
        String host = "localhost";
        int port = 9000;
        
        // Parse command line arguments
        if (args.length >= 1) {
            host = args[0];
        }
        if (args.length >= 2) {
            try {
                port = Integer.parseInt(args[1]);
            } catch (NumberFormatException e) {
                System.err.println("Invalid port number: " + args[1]);
                System.exit(1);
            }
        }
        
        AuthClient client = new AuthClient(host, port);
        if (client.connect()) {
            client.run();
        } else {
            System.err.println("Could not connect to server. Please check:");
            System.err.println("1. Server is running on " + host + ":" + port);
            System.err.println("2. Firewall settings allow the connection");
            System.err.println("3. Server host and port are correct");
            System.exit(1);
        }
    }
}

File: auth_client.py
#!/usr/bin/env python3
"""
AuthClient - Python TCP Client for User Authentication Server
Connects to Python auth server and provides interactive CLI
"""

import socket
import sys
import threading


class AuthClient:
    def __init__(self, host='localhost', port=9000):
        self.host = host
        self.port = port
        self.socket = None
        self.current_token = None
        self.running = True
    
    def connect(self):
        """Connect to the authentication server."""
        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.connect((self.host, self.port))
            print(f"✓ Connected to auth server at {self.host}:{self.port}")
            return True
        except Exception as e:
            print(f"✗ Failed to connect: {e}")
            return False
    
    def disconnect(self):
        """Disconnect from server."""
        try:
            if self.current_token:
                self.send_command(f"LOGOUT {self.current_token}")
                self.read_response()
            self.send_command("QUIT")
            self.read_response()
            self.socket.close()
        except:
            pass
        print("\nDisconnected from server.")
    
    def send_command(self, command):
        """Send command to server."""
        try:
            self.socket.sendall((command + '\n').encode('utf-8'))
            print(f"[SENT] {command}")
        except Exception as e:
            print(f"Error sending: {e}")
            raise
    
    def read_response(self):
        """Read response from server."""
        try:
            data = self.socket.recv(4096).decode('utf-8').strip()
            print(f"[RECV] {data}")
            return data
        except Exception as e:
            print(f"Error reading: {e}")
            raise
    
    def register(self):
        """Register a new user."""
        username = input("Enter username: ").strip()
        password = input("Enter password: ").strip()
        
        if not username or not password:
            print("Error: Username and password cannot be empty")
            return
        
        self.send_command(f"REGISTER {username} {password}")
        response = self.read_response()
        
        if response == "OK":
            print("✓ Registration successful!")
        else:
            print(f"✗ Registration failed: {response}")
    
    def login(self):
        """Login with credentials."""
        username = input("Enter username: ").strip()
        password = input("Enter password: ").strip()
        
        if not username or not password:
            print("Error: Username and password cannot be empty")
            return
        
        self.send_command(f"LOGIN {username} {password}")
        response = self.read_response()
        
        if response.startswith("TOKEN:"):
            self.current_token = response[6:]
            print("✓ Login successful!")
            print("✓ Session token acquired (hidden for security)")
        else:
            print(f"✗ Login failed: {response}")
    
    def access_resource(self):
        """Access a protected resource."""
        if not self.current_token:
            print("Error: You must login first")
            return
        
        resource_id = input("Enter resource ID: ").strip()
        if not resource_id:
            print("Error: Resource ID cannot be empty")
            return
        
        self.send_command(f"ACCESS {self.current_token} {resource_id}")
        response = self.read_response()
        
        if response.startswith("DATA:"):
            print(f"✓ Access granted: {response[5:]}")
        else:
            print(f"✗ Access denied: {response}")
    
    def check_status(self):
        """Check session status."""
        if not self.current_token:
            print("Error: You must login first")
            return
        
        self.send_command(f"STATUS {self.current_token}")
        response = self.read_response()
        
        if response.startswith("ACTIVE:"):
            parts = response.split(":")
            if len(parts) >= 3:
                print(f"✓ Session active for user: {parts[1]}")
                print(f"✓ Minutes remaining: {parts[2]}")
            else:
                print("✓ Session active")
        else:
            print(f"✗ Session invalid: {response}")
            self.current_token = None
    
    def logout(self):
        """Logout current user."""
        if not self.current_token:
            print("Error: No active session")
            return
        
        self.send_command(f"LOGOUT {self.current_token}")
        response = self.read_response()
        
        if response == "OK":
            print("✓ Logout successful!")
            self.current_token = None
        else:
            print(f"✗ Logout failed: {response}")
    
    def print_help(self):
        """Print help menu."""
        print("\n" + "="*50)
        print("           AVAILABLE COMMANDS")
        print("="*50)
        print("  1. register  - Register a new user")
        print("  2. login     - Login with credentials")
        print("  3. access    - Access a protected resource")
        print("  4. status    - Check session status")
        print("  5. logout    - Logout current session")
        print("  6. help      - Show this help menu")
        print("  7. quit      - Exit the client")
        print("="*50 + "\n")
    
    def run(self):
        """Main interactive loop."""
        print("\n" + "="*50)
        print("    USER AUTHENTICATION CLIENT (Python)")
        print("="*50)
        print(f"Server: {self.host}:{self.port}")
        print("="*50)
        self.print_help()
        
        while self.running:
            try:
                command = input("auth> ").strip().lower()
                
                if command in ['1', 'register', 'reg']:
                    self.register()
                elif command in ['2', 'login', 'li']:
                    self.login()
                elif command in ['3', 'access', 'acc']:
                    self.access_resource()
                elif command in ['4', 'status', 'st']:
                    self.check_status()
                elif command in ['5', 'logout', 'lo']:
                    self.logout()
                elif command in ['6', 'help', 'h', '?']:
                    self.print_help()
                elif command in ['7', 'quit', 'exit', 'q']:
                    self.running = False
                elif command == '':
                    continue
                else:
                    print(f"Unknown command: {command}")
                    print("Type 'help' for available commands")
            
            except KeyboardInterrupt:
                print("\nUse 'quit' to exit properly")
            except Exception as e:
                print(f"\nConnection error: {e}")
                print("Attempting to reconnect...")
                if not self.reconnect():
                    print("Failed to reconnect. Exiting...")
                    self.running = False
        
        self.disconnect()
        print("Goodbye!")
    
    def reconnect(self):
        """Attempt to reconnect to server."""
        try:
            self.socket.close()
        except:
            pass
        return self.connect()


def main():
    import argparse
    parser = argparse.ArgumentParser(description='Authentication Client')
    parser.add_argument('--host', default='localhost', help='Server host (default: localhost)')
    parser.add_argument('--port', type=int, default=9000, help='Server port (default: 9000)')
    args = parser.parse_args()
    
    client = AuthClient(args.host, args.port)
    if client.connect():
        client.run()
    else:
        print("\nCould not connect to server. Please check:")
        print("1. Server is running on {}:{}".format(args.host, args.port))
        print("2. Firewall settings allow the connection")
        print("3. Server host and port are correct")
        sys.exit(1)


if __name__ == "__main__":
    main()
